src/details/ArborX_DetailsDistributor.hpp

```diff
@@ -31,6 +31,103 @@ namespace ArborX
 namespace Details
 {
 
+// Assuming that batched_ranks might contain elements multiply, but duplicates
+// are not separated by other elements, compute the array of indices that sort
+// the input array (in reverse order) but also return the sorted unique elements
+// in that array with the corresponding element counts and displacement
+// (offsets).
+template <typename InputView, typename OutputView>
+static void
+determineBufferLayout(InputView batched_ranks, InputView batched_offsets,
+                      OutputView permutation_indices,
+                      std::vector<int> &unique_ranks, std::vector<int> &counts,
+                      std::vector<int> &offsets)
+{
+  auto const batched_offsets_host =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), batched_offsets);
+
+  auto const batched_ranks_host =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), batched_ranks);
+
+  ARBORX_ASSERT(unique_ranks.empty());
+  ARBORX_ASSERT(offsets.empty());
+  ARBORX_ASSERT(counts.empty());
+  ARBORX_ASSERT(permutation_indices.extent_int(0) == 0);
+  ARBORX_ASSERT(batched_ranks.size() + 1 == batched_offsets.size());
+  static_assert(
+      std::is_same<typename InputView::non_const_value_type, int>::value, "");
+  static_assert(std::is_same<typename OutputView::value_type, int>::value, "");
+
+  // In case all the batches are empty, return an empty list of unique_ranks and
+  // counts, but still have one element in offsets. This is conforming with
+  // creating the total offsets from batched_ranks and batched_offsets ignoring
+  // empty batches and calling sortAndDetermeineBufferLayout.
+  offsets.push_back(0);
+
+  auto const n = batched_ranks.size();
```

Comment:
Maybe replace `n` by `n_batched_ranks`? We have two quantities floating around (offsets and ranks), that are offset by 1, so it would be nice to specify which one it is.